File: packages/gh-tasks/src/lib/project.test.ts
import { describe, expect, it } from 'vitest';
import {
  ProjectError,
  parseProjectFlag,
  parseProjectIdentifier,
  resolveProjectRef,
} from './project.ts';

describe('parseProjectIdentifier', () => {
  it('parses owner/number', () => {
    expect(parseProjectIdentifier('ozzy-labs/5')).toEqual({ owner: 'ozzy-labs', number: 5 });
  });

  it('returns null for malformed inputs', () => {
    expect(parseProjectIdentifier(undefined)).toBeNull();
    expect(parseProjectIdentifier('')).toBeNull();
    expect(parseProjectIdentifier('ozzy-labs')).toBeNull();
    expect(parseProjectIdentifier('/5')).toBeNull();
    expect(parseProjectIdentifier('ozzy-labs/')).toBeNull();
    expect(parseProjectIdentifier('ozzy-labs/abc')).toBeNull();
    expect(parseProjectIdentifier('ozzy-labs/0')).toBeNull();
    expect(parseProjectIdentifier('ozzy-labs/-1')).toBeNull();
    expect(parseProjectIdentifier('ozzy-labs/3.5')).toBeNull();
  });

  it('trims surrounding whitespace', () => {
    expect(parseProjectIdentifier('  ozzy-labs/5  ')).toEqual({ owner: 'ozzy-labs', number: 5 });
  });
});

describe('parseProjectFlag', () => {
  it('returns null when no --project flag is present', () => {
    expect(parseProjectFlag([])).toBeNull();
    expect(parseProjectFlag(['add', 'foo'])).toBeNull();
  });

  it('parses --project=<owner>/<number>', () => {
    expect(parseProjectFlag(['--project=ozzy-labs/5'])).toEqual({
      owner: 'ozzy-labs',
      number: 5,
    });
  });

  it('parses --project <owner>/<number> (separate arg)', () => {
    expect(parseProjectFlag(['--project', 'ozzy-labs/5'])).toEqual({
      owner: 'ozzy-labs',
      number: 5,
    });
  });

  it('throws on missing value', () => {
    expect(() => parseProjectFlag(['--project'])).toThrow(ProjectError);
  });

  it('throws on malformed value', () => {
    expect(() => parseProjectFlag(['--project=bogus'])).toThrow(ProjectError);
    expect(() => parseProjectFlag(['--project=ozzy-labs/0'])).toThrow(ProjectError);
  });
});

describe('resolveProjectRef', () => {
  it('returns the --project flag value when present', () => {
    expect(resolveProjectRef({ scope: 'org', argv: ['--project=ozzy-labs/5'] })).toEqual({
      owner: 'ozzy-labs',
      number: 5,
    });
  });

  it('falls back to config.orgProject for org scope', () => {
    expect(
      resolveProjectRef({
        scope: 'org',
        argv: [],
        config: { orgProject: { owner: 'ozzy-labs', number: 7 } },
      })
    ).toEqual({ owner: 'ozzy-labs', number: 7 });
  });

  it('falls back to config.userProject for user scope', () => {
    expect(
      resolveProjectRef({
        scope: 'user',
        argv: [],
        config: { userProject: { owner: 'ozzy-3', number: 2 } },
      })
    ).toEqual({ owner: 'ozzy-3', number: 2 });
  });

  it('flag outranks config', () => {
    expect(
      resolveProjectRef({
        scope: 'org',
        argv: ['--project=ozzy-labs/5'],
        config: { orgProject: { owner: 'other', number: 99 } },
      })
    ).toEqual({ owner: 'ozzy-labs', number: 5 });
  });

  it('does not cross-pollinate org / user config keys', () => {
    expect(() =>
      resolveProjectRef({
        scope: 'user',
        argv: [],
        config: { orgProject: { owner: 'ozzy-labs', number: 5 } },
      })
    ).toThrow(ProjectError);
    expect(() =>
      resolveProjectRef({
        scope: 'org',
        argv: [],
        config: { userProject: { owner: 'ozzy-3', number: 2 } },
      })
    ).toThrow(ProjectError);
  });

  it('throws when neither flag nor config is set', () => {
    expect(() => resolveProjectRef({ scope: 'org', argv: [] })).toThrow(ProjectError);
    expect(() => resolveProjectRef({ scope: 'user', argv: [] })).toThrow(ProjectError);
  });

  it('throws when called with repo scope', () => {
    expect(() => resolveProjectRef({ scope: 'repo', argv: ['--project=ozzy-labs/5'] })).toThrow(
      ProjectError
    );
  });
});

File: internal/project/project_test.go
package project_test

import (
	"testing"

	"github.com/ozzy-labs/gh-tasks/internal/project"
	"github.com/ozzy-labs/gh-tasks/internal/scope"
)

func TestParseIdentifier(t *testing.T) {
	t.Parallel()

	cases := []struct {
		in   string
		want project.Ref
		ok   bool
	}{
		{"ozzy-labs/3", project.Ref{Owner: "ozzy-labs", Number: 3}, true},
		{"a/1", project.Ref{Owner: "a", Number: 1}, true},
		{"  ozzy-labs/12  ", project.Ref{Owner: "ozzy-labs", Number: 12}, true},
		{"ozzy-labs/abc", project.Ref{}, false},
		{"ozzy-labs/0", project.Ref{}, false},
		{"ozzy-labs/-1", project.Ref{}, false},
		{"ozzy-labs/", project.Ref{}, false},
		{"/3", project.Ref{}, false},
		{"", project.Ref{}, false},
	}
	for _, tc := range cases {
		t.Run(tc.in, func(t *testing.T) {
			t.Parallel()
			got, ok := project.ParseIdentifier(tc.in)
			if ok != tc.ok || got != tc.want {
				t.Errorf("got (%v,%v) want (%v,%v)", got, ok, tc.want, tc.ok)
			}
		})
	}
}

func TestParseFlag(t *testing.T) {
	t.Parallel()

	t.Run("absent", func(t *testing.T) {
		t.Parallel()
		_, present, err := project.ParseFlag([]string{"--scope=org"})
		if err != nil || present {
			t.Fatalf("got (present=%v, err=%v)", present, err)
		}
	})

	t.Run("equals-form", func(t *testing.T) {
		t.Parallel()
		ref, present, err := project.ParseFlag([]string{"--project=foo/7"})
		if err != nil || !present {
			t.Fatalf("got (present=%v, err=%v)", present, err)
		}
		if ref.Owner != "foo" || ref.Number != 7 {
			t.Errorf("got %v", ref)
		}
	})

	t.Run("missing-value-is-present-with-error", func(t *testing.T) {
		t.Parallel()
		_, present, err := project.ParseFlag([]string{"--project"})
		if err == nil {
			t.Fatalf("want error")
		}
		if !present {
			t.Errorf("present=false on malformed flag; want true (flag was present)")
		}
	})

	t.Run("invalid-identifier-is-present-with-error", func(t *testing.T) {
		t.Parallel()
		_, present, err := project.ParseFlag([]string{"--project=bad"})
		if err == nil {
			t.Fatalf("want error")
		}
		if !present {
			t.Errorf("present=false on malformed flag; want true (flag was present)")
		}
	})
}

func TestResolve(t *testing.T) {
	t.Parallel()

	t.Run("repo-scope-errors", func(t *testing.T) {
		t.Parallel()
		_, err := project.Resolve(project.ResolveOptions{Scope: scope.Repo})
		if err == nil {
			t.Fatalf("want error")
		}
	})

	t.Run("flag-wins", func(t *testing.T) {
		t.Parallel()
		got, err := project.Resolve(project.ResolveOptions{
			Scope:      scope.Org,
			Argv:       []string{"--project=foo/7"},
			OrgProject: project.Ref{Owner: "config", Number: 9},
		})
		if err != nil {
			t.Fatalf("err: %v", err)
		}
		want := project.Ref{Owner: "foo", Number: 7}
		if got != want {
			t.Errorf("got %v want %v", got, want)
		}
	})

	t.Run("falls-back-to-org-config", func(t *testing.T) {
		t.Parallel()
		got, err := project.Resolve(project.ResolveOptions{
			Scope:      scope.Org,
			OrgProject: project.Ref{Owner: "configured", Number: 5},
		})
		if err != nil {
			t.Fatalf("err: %v", err)
		}
		if got.Number != 5 || got.Owner != "configured" {
			t.Errorf("got %v", got)
		}
	})

	t.Run("falls-back-to-user-config", func(t *testing.T) {
		t.Parallel()
		got, err := project.Resolve(project.ResolveOptions{
			Scope:       scope.User,
			UserProject: project.Ref{Owner: "u", Number: 1},
		})
		if err != nil {
			t.Fatalf("err: %v", err)
		}
		if got.Owner != "u" {
			t.Errorf("got %v", got)
		}
	})

	t.Run("no-flag-no-config-errors", func(t *testing.T) {
		t.Parallel()
		_, err := project.Resolve(project.ResolveOptions{Scope: scope.Org})
		if err == nil {
			t.Fatalf("want error")
		}
	})
}
